carcara/src/parser/rare/mod.rs

```diff
@@ -0,0 +1,156 @@
+enum Body {
+    Conclusion(Rc<Term>),
+    Premise(Vec<Rc<Term>>),
+    Args(Vec<String>),
+}
+
+fn parse_parameters<R: BufRead>(parser: &mut Parser<R>) -> CarcaraResult<(String, TypeParameter)> {
+    parser.expect_token(Token::OpenParen)?;
+    let name = parser.expect_symbol()?;
+    let term = parser.parse_sort(true)?;
+
+    parser.insert_sorted_var((name.clone(), term.clone()));
+    parser.state.sort_defs.insert(
+        name.clone(),
+        SortDef {
+            body: term.clone(),
+            params: Vec::default(),
+        },
+    );
+    let current_token = &parser.current_token;
+    match current_token {
+        Token::CloseParen => {
+            parser.expect_token(Token::CloseParen)?;
+            Ok((
+                name,
+                TypeParameter {
+                    term,
+                    attribute: AttributeParameters::None,
+                },
+            ))
+        }
+        Token::Keyword(_) => {
+            let kind_of_arg = parser.expect_keyword()?;
+            parser.expect_token(Token::CloseParen)?;
+            if kind_of_arg == "list" {
+                return Ok((
+                    name,
+                    TypeParameter {
+                        term,
+                        attribute: AttributeParameters::List,
+                    },
+                ));
+            }
+            Err(Error::Parser(
+                ParserError::InvalidRareArgAttribute(kind_of_arg),
+                parser.current_position,
+            ))
+        }
+        _ => Err(Error::Parser(
+            ParserError::UnexpectedToken(current_token.clone()),
+            parser.current_position,
+        )),
+    }
+}
+
+fn parse_body<R: BufRead>(parser: &mut Parser<R>) -> CarcaraResult<Body> {
+    let qualified_arg: Vec<char> = parser.expect_keyword()?.chars().collect();
+    match qualified_arg.as_slice() {
+        ['c', 'o', 'n', 'c', 'l', 'u', 's', 'i', 'o', 'n', ..] => {
+            let rewrite_term = parser.parse_term()?;
+            Ok(Body::Conclusion(rewrite_term))
+        }
+        ['a', 'r', 'g', 's', ..] => {
+            fn parse_args<R: BufRead>(parser: &mut Parser<R>) -> CarcaraResult<Vec<String>> {
+                parser.expect_token(Token::OpenParen)?;
+                parser.parse_sequence(super::Parser::expect_symbol, false)
+            }
+            let args = parse_args(parser)?;
+            Ok(Body::Args(args))
+        }
+        ['p', 'r', 'e', 'm', 'i', 's', 'e', 's', ..] => {
+            parser.expect_token(Token::OpenParen)?;
+            let terms = parser.parse_sequence(
+                |parser| {
+                    let term = parser.parse_term()?;
+                    Ok(term)
+                },
+                false,
+            )?;
+            Ok(Body::Premise(terms))
+        }
+        attribute => Err(Error::Parser(
+            ParserError::InvalidRareFunctionAttribute(attribute.iter().collect()),
+            parser.current_position,
+        )),
+    }
```

Comment:
I think this can be simplified to:
```suggestion
    let qualified_arg = parser.expect_keyword()?;
    match qualified_arg.as_str() {
        "conclusion" => {
            let rewrite_term = parser.parse_term()?;
            Ok(Body::Conclusion(rewrite_term))
        }
        "args" => {
            fn parse_args<R: BufRead>(parser: &mut Parser<R>) -> CarcaraResult<Vec<String>> {
                parser.expect_token(Token::OpenParen)?;
                parser.parse_sequence(super::Parser::expect_symbol, false)
            }
            let args = parse_args(parser)?;
            Ok(Body::Args(args))
        }
        "premises" => {
            parser.expect_token(Token::OpenParen)?;
            let terms = parser.parse_sequence(
                |parser| {
                    let term = parser.parse_term()?;
                    Ok(term)
                },
                false,
            )?;
            Ok(Body::Premise(terms))
        }
        _ => Err(Error::Parser(
            ParserError::InvalidRareFunctionAttribute(qualified_arg),
            parser.current_position,
        )),
    }
```